SI-9102: Reflect method invoke with mixed args

som-snytt opened the pull request:
A missing default branch when a method had value class
or by-name params caused other args to present as null
under reflective invocation.

Review by @xeno-by

som-snytt committed:
SI-9102: Reflect method invoke with mixed args

A missing default branch when a method had value class
or by-name params caused other args to present as null
under reflective invocation.

xeno-by commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -428,9 +428,12 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
+            else if (isByName(i))      () => arg // don't transform by-name value class params
+            else if (isDerivedValueClass(i)) paramUnboxers(i).invoke(arg)
+            else                             arg
+          )
```
Oops, yeah, I missed this. Thanks for the fix!

som-snytt replied:
It surfaced on scala-user. I'm not sure how, the neighbor code is not new. Are people just not doing enough reflective invocation?

xeno-by commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -428,9 +428,12 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
-          else if (isDerivedValueClass(i)) args1(i) = paramUnboxers(i).invoke(arg)
+          args1(i) = (
+            if (i >= paramCount)             arg // don't transform varargs
+            else if (isByName(i))      () => arg // don't transform by-name value class params
```
I think that formatting here is somewhat misleading.

som-snytt replied:
You mean only the lonely parens? I know the whole block is a bit over-reaching for new heights of clarity. I can't read if/else anymore since reading Scala pattern matches with the arrows lining up. Let me know if pulling the if/else to RHS is over the top.

xeno-by commented:
LGTM

som-snytt committed:
SI-9102: Improve test

Cover the second use case reported on the ML (ctors).

Improve formatting per the review. And it really does look
a lot better.

retronym commented:
LGTM, too.
